pipelinerl/launch.py

```diff
@@ -461,7 +462,13 @@ def gently_stop_all_processes():
                 logger.info(f"Process {proc.handle.args} finished cleanly")
                 alive.remove(proc)
             if alive and all(is_inference_process(proc) for proc in alive):
-                logger.info(f"All pipeline workers finished; stopping {len(alive)} inference server(s)")
+                # shut down inference servers after training is complete
+                if trainer_state is not None and not trainer_state.training_done:
+                    # check if training is completed
+                    logger.info(f"Waiting for training completion signal (training_done={trainer_state.training_done})")
```

Comment:
we only can get in this branch if all the processes in `alive` list seems to be inference ones. Is it correct to expect here for trainer to send us something? I assume that the trainer is dead already at that moment.

Reply:
yes, this is the point of what's happening in this pr, the trainer will flip the training_done switch when it's reached the last training step. at that point we should only have inference servers alive so we can proceed to shut them down.